rand_distr/src/utils.rs

```diff
@@ -10,8 +10,50 @@
 
 use rand::Rng;
 use crate::ziggurat_tables;
-
 use rand::distributions::hidden_export::IntoFloat;
+use core::{cmp, ops};
+
+/// Trait for floating-point scalar types
+/// 
+/// This allows many distributions to work with `f32` or `f64` parameters and is
+/// potentially extensible. Note however that the `Exp1` and `StandardNormal`
+/// distributions are implemented exclusively for `f32` and `f64`.
+/// 
+/// The bounds and methods are based purely on internal
+/// requirements, and will change as needed.
+pub trait Float: Copy + Sized + cmp::PartialOrd
+    + ops::Add<Output = Self>
+    + ops::Sub<Output = Self>
+    + ops::Mul<Output = Self>
+    + ops::Div<Output = Self>
+{
+    /// Support approximate representation of a f64 value
+    fn from(x: f64) -> Self;
+    /// Take the exponential of self
+    fn exp(self) -> Self;
+    /// Take the natural logarithm of self
+    fn ln(self) -> Self;
+    /// Take square root of self
+    fn sqrt(self) -> Self;
+    /// Take self to a floating-point power
+    fn powf(self, power: Self) -> Self;
+}
+
+impl Float for f32 {
+    fn from(x: f64) -> Self { x as f32 }
```

Comment:
Can we have undefined behavior when converting from a large `f64` to `f32`?

Reply:
From memory, I don't think it's UB. [Example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1a4c633980a699c624d6a863ff267b4a).

Reply:
No, [you're right](https://doc.rust-lang.org/reference/expressions/operator-expr.html#semantics) (unless the spec isn't up to date with the compiler).

There is one case of `N::from(1. / 3.)` so I'd prefer not to restrict precision to `f32` — unless we rewrite this as `N::from(1.) / N::from(3.)`.

Reply:
Correction: [it's not UB](https://github.com/rust-lang/rust/issues/15536), however the reference is out of date.